### PR TITLE
feat: borrow loans

### PR DESCRIPTION
**Pull request by vbstreetz**
🚧 Very much a work in progress:
- Adds /loans/new and /loans/list routes:
  - /loans/new: borrow form
  - /loans/list: lists the active user borrows

### CODE REVIEW

File: contracts/erc20.js
Comment (evgenyboxer) [inline contracts/erc20.js:1]: could this be a TS file instead?

File: queries/loans/types.ts
Comment (evgenyboxer) [inline queries/loans/types.ts:1]: Could we use our BigNumber library instead?

File: sections/loans/components/ActionBox/ActionBox.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/ActionBox.tsx:29]: for the code above ^ you could use `useMemo` here no?

File: sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx:15]: minor `string[]` :)

File: sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx:69]: please put this into the translation file

File: sections/loans/components/ActionBox/BorrowSynthsTab/AssetInput.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/BorrowSynthsTab/AssetInput.tsx:29]: can probably `useMemo` yeah?

File: utils/transactions.ts
Comment (evgenyboxer) [inline utils/transactions.ts:25]: @vbstreetz could you explain what is `callStatic`?

File: utils/formatters/big-number.ts
Comment (evgenyboxer) [inline utils/formatters/big-number.ts:1]: can we use `BigNumber.js`?

File: contracts/erc20.js
Comment (vbstreetz) [reply contracts/erc20.js:1]: Awesome. On it..

File: sections/loans/components/ActionBox/ActionBox.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/ActionBox.tsx:29]: Sure.

File: sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/ActiveBorrowsTab/LoanList.tsx:71]: Awesome.

File: utils/transactions.ts
Comment (vbstreetz) [reply utils/transactions.ts:25]: Oh, this is a dry run of a write transaction. It's basically a hack to retrieve the revert message of a failed transaction.

File: queries/loans/types.ts
Comment (vbstreetz) [reply queries/loans/types.ts:1]: Switched. Wish bignumber.js had the same API as ethers.BigNumber, similar to big.js though.

File: translations/en.json
Comment (evgenyboxer) [inline translations/en.json:598]: Hey @vbstreetz we usually don't uppercase translations as this might confuse translators. You can use `uppercase` via css for such situations.

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/components/Balance.tsx:128]: is this needed?

File: translations/en.json
Comment (evgenyboxer) [inline translations/en.json:540]: This is just a placeholder for now?

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (evgenyboxer) [inline sections/loans/components/ActionBox/components/Balance.tsx:157]: the "MAX" text needs to be translated

File: utils/transactions.ts
Comment (evgenyboxer) [inline utils/transactions.ts:26]: great work with this function!

File: sections/loans/components/InfoBox/InfoBox.tsx
Comment (evgenyboxer) [inline sections/loans/components/InfoBox/InfoBox.tsx:71]: could you use `Synths.sBTC`, etc for this?

File: translations/en.json
Comment (vbstreetz) [reply translations/en.json:598]: Awesome. On it..

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/components/Balance.tsx:128]: You mean the sleep? I'll remote it..

**Review by evgenyboxer — APPROVED** [review]

**Comment by evgenyboxer** [comment]
@vbstreetz could you please fix the conflicts :) 
This is now ready to merge!

**Comment by evgenyboxer** [comment]
@vbstreetz still has some conflicting files, do you need help with this?

**Comment by vbstreetz** [comment]
> @vbstreetz still has some conflicting files, do you need help with this?

Gotcha. It's just the package.json lock file that keeps being updated by other merges. Should be good to merge once this is approved.

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanActions/Wrapper.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanActions/Wrapper.tsx:3]: Please use date-fns. Not a big fan of momentjs

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx:49]: Probably needs to be in the translation file

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx:78]: You can probably just target body? What do you think? Then we don't need to add the `div` at the root.

File: sections/loans/components/ActionBox/BorrowSynthsTab/FormButton.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/BorrowSynthsTab/FormButton.tsx:39]: Love this thing

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/components/Balance.tsx:109]: Not sure I get this. Why are you using `isMounted` exactly?

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/components/Balance.tsx:131]: Events are really bad with ethers.js. They keep polling the infura node, every 10 secs or so.Any way we can use the graph for this?

Comment (clementbalestrat) [reply sections/loans/components/ActionBox/components/Balance.tsx:131]: Happy to help you setting this up if needed

File: sections/loans/components/InfoBox/InfoBox.tsx
Comment (clementbalestrat) [inline sections/loans/components/InfoBox/InfoBox.tsx:90]: calling `load()` on every block might be a bit aggressive. it's every 10-15secs. I think we should really avoid this if possible. Do you think it's possible?

**Review by clementbalestrat — CHANGES_REQUESTED** [review]
Awesome stuff. A few minors comments/open questions.

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (clementbalestrat) [inline sections/loans/components/ActionBox/components/Balance.tsx:138]: Amazing

**Review by clementbalestrat — CHANGES_REQUESTED** [review]

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (clementbalestrat) [reply sections/loans/components/ActionBox/components/Balance.tsx:138]: I made a comment earlier (deleted) about etherjs events. But the fact that you unsubscribe when the component is unmounted is great.

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (clementbalestrat) [reply sections/loans/components/ActionBox/components/Balance.tsx:109]: All good @evgenyboxer just explained it to me. Though I wonder if React has a "native" solution for that.

File: Makefile
Comment (vbstreetz) [reply Makefile:1]: Hey. I'll remove it here in favor of manual npm run 👍🏽

File: utils/transactions.ts
Comment (vbstreetz) [reply utils/transactions.ts:26]: Awesome. Hoping it gets considered to be merged in https://github.com/Synthetixio/js-monorepo/tree/master/packages/transaction-notifier @clementbalestrat  ..

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/components/Balance.tsx:109]: Gotcha. There's been talk of a `useAsyncEffect` i think?

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/components/Balance.tsx:138]: Awesome. Maybe @dvd-schwrtz knows if [subscriptions for subgraphs](https://thegraph.com/docs/graphql-api#schema) is being worked on soon, so we don't have to subscribe to the events directly?

Comment (clementbalestrat) [reply sections/loans/components/ActionBox/components/Balance.tsx:138]: Yeah that would be great!

File: utils/transactions.ts
Comment (clementbalestrat) [reply utils/transactions.ts:15]: We should probably catch up on this, to see if we want to use it everywhere in the dApp

**Comment by clementbalestrat** [comment]
@vbstreetz I just pushed some fixes to make sure it is mergeable with `dev`.

File: utils/transactions.ts
Comment (vbstreetz) [reply utils/transactions.ts:15]: Awesome

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx]: Done.

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanActions/Wrapper.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanActions/Wrapper.tsx]: Done 👍🏽

File: sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx
Comment (vbstreetz) [reply sections/loans/components/ActionBox/ActiveBorrowsTab/ModifyLoanMenu.tsx:80]: Gotcha. I've actually switched this to use tippy..

File: sections/loans/components/ActionBox/components/Balance.tsx
Comment (evgenyboxer) [reply sections/loans/components/ActionBox/components/Balance.tsx:109]: good call, we could use either that or a custom hook that does the check for us.

**Comment by vbstreetz** [comment]
I think we can merge this then work on https://github.com/Synthetixio/staking/issues/435 in a separate ticket.

**Review by JChiaramonte7 — APPROVED** [review]
Ran through the QA checklist here #450 
- notable issues with claim button and numeric input were fixed

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:13]: is `it.only` on purpose here?

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:17]: looks like a preparation for testing, should be placed inside `before()`

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:20]: it should be a separate test defined with `it('something')`

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:28]: it should be a separate test defined with `it('something')`

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:38]: it should be a separate test defined with `it('something')`

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:48]: it should be a separate test defined with `it('something')`

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [inline tests/e2e/specs/loans-spec.js:6]: empty `before()`; it's useless

File: tests/e2e/specs/loans-spec.js
Comment (drptbl) [reply tests/e2e/specs/loans-spec.js:13]: it should be separated to two tests; you can use `context()` for example (or use different`*-spec.js` files and do the preparation for manage loan test inside `before()`):
1. should create loan
2. should manage loan? (needs more descriptive name)
```
describe('Loans tests', () => {
	before(() => {
	...
	});
	context('Create a loan', () => {
		it(`should ..`, () => {
		});
		it(`should ..`, () => {
		});
		...
	});
	context('Manage a loan', () => {
		it(`should ..`, () => {
		});
		it(`should ..`, () => {
		});
		...
	});
});
```

Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js:13]: This one as well. Best to run it in one test case to speed up testing.

Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Did split the test into smaller cases.

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Done

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Done

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Done

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Done

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js:38]: Did split the test into smaller cases.

File: tests/e2e/specs/loans-spec.js
Comment (vbstreetz) [reply tests/e2e/specs/loans-spec.js]: Done

File: queries/withdrawals/useGetWithdrawalsDataQuery.ts
Comment (clementbalestrat) [inline queries/withdrawals/useGetWithdrawalsDataQuery.ts:4]: what is this comment? is it still needed?

**Review by clementbalestrat — APPROVED** [review]
@vbstreetz solid work man, really.

File: queries/withdrawals/useGetWithdrawalsDataQuery.ts
Comment (vbstreetz) [reply queries/withdrawals/useGetWithdrawalsDataQuery.ts:4]: Issue seems to still be present. See next line.

Comment (vbstreetz) [reply queries/withdrawals/useGetWithdrawalsDataQuery.ts:4]: I'll merge and resolve in a new PR if issue goes away.

**Pull request merged**
